File: src/java/LRUCache.java
package java;

import java.util.HashMap;
import java.util.Map;

class LRUCache {

    private Node head;
    private Node tail;
    private int capacity;
    private int size;
    private Map<Integer, Node> cache = new HashMap<>();

    public LRUCache(int capacity) {
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.pre = head;
        this.capacity = capacity;
    }

    public int get(int key) {
        if (cache.containsKey(key)) {
            Node node = cache.get(key);
            moveToHead(node);
            return node.value;
        } else {
            return -1;
        }
    }

    private void moveToHead(Node node) {
        node.pre.next = node.next;
        node.next.pre = node.pre;
        addToHead(node);

    }

    public void put(int key, int value) {

        if (cache.containsKey(key)) {
            Node node = cache.get(key);
            node.value = value;
            moveToHead(node);
        } else {
            Node node = new Node(key, value);
            addToHead(node);
            cache.put(key, node);
            if (++size > capacity) {
                deleteTail();
            }
        }
    }

    private void addToHead(Node node) {
        Node next = head.next;
        head.next = node;
        node.pre = head;
        next.pre = node;
        node.next = next;
    }

    private void deleteTail() {

        Node pre = tail.pre;

        pre.pre.next = tail;
        tail.pre = pre.pre;


        cache.remove(pre.key);
        size--;


    }


    class Node {

        int key;
        int value;
        Node next;
        Node pre;

        public Node(int key, int value) {
            this.key = key;
            this.value = value;
        }

        public Node() {
        }
    }
}

File: src/design/LRUCache.kt
package design


class LRUCache(capacity: Int) {
    var capacity = capacity

    var map = mutableMapOf<Int, Node>()
    var head = Node(0, 0)
    var tail = Node(0, 0)
    var size = 0

    init {
        head.next = tail
        tail.pre = head
    }

    fun get(key: Int): Int {
        val node = map[key] ?: return -1
        moveToHead(node)
        return node.value
    }

    fun put(key: Int, value: Int) {
        var cache = map[key]
        if (cache != null) {
            cache.value = value
            moveToHead(cache)
        } else {
            val node = Node(value, key)
            map[key] = node
            addToHead(node)
            if (++size > capacity) {
                var node = tail.pre!!
                deleteTail(node)
                map.remove(node.key)
                size--
            }
        }

    }

    private fun deleteTail(node: Node) {
        node.pre?.next = node.next
        node.next?.pre = node.pre
    }

    private fun addToHead(node: Node) {
        node.next = head.next
        node.pre = head
        head.next?.pre = node
        head.next = node
    }

    private fun moveToHead(node: Node) {
        deleteTail(node)
        addToHead(node)

    }

    class Node(var value: Int, var key: Int) {

        var pre: Node? = null
        var next: Node? = null


    }

}